{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare RNA-seq counts of intergenic intervals that overlap with an ATAC-seq peak vs. those that don't"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load peaks data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromsome</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>peak_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100027916</td>\n",
       "      <td>100029543</td>\n",
       "      <td>Peak_162534,Peak_189829,Peak_205380,Peak_23225...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100036789</td>\n",
       "      <td>100037403</td>\n",
       "      <td>Peak_133803,Peak_173137,Peak_67659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100037554</td>\n",
       "      <td>100039283</td>\n",
       "      <td>Peak_144687,Peak_179877,Peak_39266,Peak_44011,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100048434</td>\n",
       "      <td>100048996</td>\n",
       "      <td>Peak_189438,Peak_222062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100049436</td>\n",
       "      <td>100049815</td>\n",
       "      <td>Peak_175134,Peak_199830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135845</th>\n",
       "      <td>chrX</td>\n",
       "      <td>9995829</td>\n",
       "      <td>9996095</td>\n",
       "      <td>Peak_69685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135846</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11097622</td>\n",
       "      <td>11097859</td>\n",
       "      <td>Peak_207213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135847</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11098228</td>\n",
       "      <td>11099027</td>\n",
       "      <td>Peak_119573,Peak_267499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135848</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12394970</td>\n",
       "      <td>12395294</td>\n",
       "      <td>Peak_227283,Peak_267500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135849</th>\n",
       "      <td>chrY</td>\n",
       "      <td>4030730</td>\n",
       "      <td>4031043</td>\n",
       "      <td>Peak_115360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>135850 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromsome      begin        end  \\\n",
       "0           chr1  100027916  100029543   \n",
       "1           chr1  100036789  100037403   \n",
       "2           chr1  100037554  100039283   \n",
       "3           chr1  100048434  100048996   \n",
       "4           chr1  100049436  100049815   \n",
       "...          ...        ...        ...   \n",
       "135845      chrX    9995829    9996095   \n",
       "135846      chrY   11097622   11097859   \n",
       "135847      chrY   11098228   11099027   \n",
       "135848      chrY   12394970   12395294   \n",
       "135849      chrY    4030730    4031043   \n",
       "\n",
       "                                                 peak_ids  \n",
       "0       Peak_162534,Peak_189829,Peak_205380,Peak_23225...  \n",
       "1                      Peak_133803,Peak_173137,Peak_67659  \n",
       "2       Peak_144687,Peak_179877,Peak_39266,Peak_44011,...  \n",
       "3                                 Peak_189438,Peak_222062  \n",
       "4                                 Peak_175134,Peak_199830  \n",
       "...                                                   ...  \n",
       "135845                                         Peak_69685  \n",
       "135846                                        Peak_207213  \n",
       "135847                            Peak_119573,Peak_267499  \n",
       "135848                            Peak_227283,Peak_267500  \n",
       "135849                                        Peak_115360  \n",
       "\n",
       "[135850 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_peaks_df = pd.read_csv('ENCSR885DVH/ENCFF879FSD.bed', sep='\\t', header=None)\n",
    "\n",
    "\n",
    "# For some reason (which I need to look into), there are multiple peaks\n",
    "# with the same start and end coordinates. I need to collapse these entries\n",
    "# down into a single interval.\n",
    "interval_to_ids = defaultdict(lambda: [])\n",
    "for r_i, row in raw_peaks_df.iterrows():\n",
    "    chrm = row[0]\n",
    "    begin = row[1]\n",
    "    end = row[2]\n",
    "    peak_id = row[3]\n",
    "    interval_to_ids[(chrm, begin, end)].append(peak_id)\n",
    "da = [\n",
    "    (k[0], k[1], k[2], ','.join(v))\n",
    "    for k, v in interval_to_ids.items()\n",
    "]\n",
    "peaks_df = pd.DataFrame(\n",
    "    data=da,\n",
    "    columns=[\n",
    "        'chromsome', \n",
    "        'begin', \n",
    "        'end', \n",
    "        'peak_ids'\n",
    "    ]\n",
    ")\n",
    "peaks_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load intergenic RNA-seq interval counts data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interval_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>interval_begin</th>\n",
       "      <th>interval_end</th>\n",
       "      <th>n_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1.1000_1200</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>1200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1.1200_1400</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1200</td>\n",
       "      <td>1400</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1.1400_1600</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1400</td>\n",
       "      <td>1600</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1.1600_1800</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1600</td>\n",
       "      <td>1800</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1.1800_2000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1800</td>\n",
       "      <td>2000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193533</th>\n",
       "      <td>chrY.57182169_57182369</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57182169</td>\n",
       "      <td>57182369</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193534</th>\n",
       "      <td>chrY.57182369_57182569</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57182369</td>\n",
       "      <td>57182569</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193535</th>\n",
       "      <td>chrY.57182569_57182769</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57182569</td>\n",
       "      <td>57182769</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193536</th>\n",
       "      <td>chrY.57182769_57182969</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57182769</td>\n",
       "      <td>57182969</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193537</th>\n",
       "      <td>chrY.57182969_57183100</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57182969</td>\n",
       "      <td>57183100</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7193538 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    interval_id chromosome  interval_begin  interval_end  \\\n",
       "0                chr1.1000_1200       chr1            1000          1200   \n",
       "1                chr1.1200_1400       chr1            1200          1400   \n",
       "2                chr1.1400_1600       chr1            1400          1600   \n",
       "3                chr1.1600_1800       chr1            1600          1800   \n",
       "4                chr1.1800_2000       chr1            1800          2000   \n",
       "...                         ...        ...             ...           ...   \n",
       "7193533  chrY.57182169_57182369       chrY        57182169      57182369   \n",
       "7193534  chrY.57182369_57182569       chrY        57182369      57182569   \n",
       "7193535  chrY.57182569_57182769       chrY        57182569      57182769   \n",
       "7193536  chrY.57182769_57182969       chrY        57182769      57182969   \n",
       "7193537  chrY.57182969_57183100       chrY        57182969      57183100   \n",
       "\n",
       "         n_reads  \n",
       "0              0  \n",
       "1              0  \n",
       "2              0  \n",
       "3              0  \n",
       "4              0  \n",
       "...          ...  \n",
       "7193533        0  \n",
       "7193534        0  \n",
       "7193535        0  \n",
       "7193536        0  \n",
       "7193537        0  \n",
       "\n",
       "[7193538 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq_df = pd.read_csv('output/ENCSR885DVH_ENCFF625ZBS.interval_counts.200_bp.ORIG.tsv', sep='\\t')\n",
    "rnaseq_df "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compute which intervals overlap with an ATAC-seq peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chromosome chrX\n",
      "Processing chromosome chr20\n",
      "Processing chromosome chr12\n",
      "Processing chromosome chr16\n",
      "Processing chromosome chr6\n",
      "Processing chromosome chr8\n",
      "Processing chromosome chr19\n",
      "Processing chromosome chr10\n",
      "Processing chromosome chr18\n",
      "Processing chromosome chr11\n",
      "Processing chromosome chr7\n",
      "Processing chromosome chr13\n",
      "Processing chromosome chr15\n",
      "Processing chromosome chr2\n",
      "Processing chromosome chrY\n",
      "Processing chromosome chr3\n",
      "Processing chromosome chr9\n",
      "Processing chromosome chr14\n",
      "Processing chromosome chr22\n",
      "Processing chromosome chr17\n",
      "Processing chromosome chr4\n",
      "Processing chromosome chr21\n",
      "Processing chromosome chr1\n",
      "Processing chromosome chr5\n",
      "159890 intervals in a peak\n",
      "7033648 intervals outside a peak\n"
     ]
    }
   ],
   "source": [
    "chrm_to_peaks = {\n",
    "    chrm: peaks_df.loc[peaks_df['chromsome'] == chrm]\n",
    "    for chrm in set(peaks_df['chromsome'])\n",
    "}\n",
    "chrm_to_peaks = {\n",
    "    chrm: peaks_df.sort_values(by='begin')\n",
    "    for chrm, peaks_df in chrm_to_peaks.items()\n",
    "}\n",
    "chrm_to_peak_begins = {\n",
    "    chrm: list(df['begin'])\n",
    "    for chrm, df in chrm_to_peaks.items()\n",
    "}\n",
    "chrm_to_peak_ends = {\n",
    "    chrm: list(df['end'])\n",
    "    for chrm, df in chrm_to_peaks.items()\n",
    "}\n",
    "\n",
    "# Set to True to output messages\n",
    "debug = False\n",
    "\n",
    "ints_in_peaks = []\n",
    "ints_outside_peaks = []\n",
    "for chrm, peak_begins in chrm_to_peak_begins.items():\n",
    "    print(\"Processing chromosome {}\".format(chrm))\n",
    "    peak_ends = chrm_to_peak_ends[chrm]\n",
    "    chrm_rnaseq_df = rnaseq_df.loc[rnaseq_df['chromosome'] == chrm]\n",
    "    chrm_rnaseq_df = chrm_rnaseq_df.sort_values(by='interval_begin')\n",
    "    interval_begins = list(chrm_rnaseq_df['interval_begin'])\n",
    "    interval_ends = list(chrm_rnaseq_df['interval_end'])\n",
    "    peak_i = 0\n",
    "        \n",
    "    for int_i, (_, row) in enumerate(chrm_rnaseq_df.iterrows()):\n",
    "        interval_id = row['interval_id']\n",
    "        int_begin = interval_begins[int_i]\n",
    "        int_end = interval_ends[int_i]\n",
    "        \n",
    "        if int_begin >= peak_begins[peak_i]:\n",
    "            # The interval beginning has crossed over \n",
    "            # a peak's beginning\n",
    "            if int_begin > peak_ends[peak_i]:\n",
    "                #   The interval is now passed the peak:\n",
    "                # \n",
    "                #   interval: ---------[------]--\n",
    "                #   peak:     --[---]------------\n",
    "                if debug:\n",
    "                    print('Interval ({}, {}) has now passed current peak ({}, {})'.format(\n",
    "                        int_begin,\n",
    "                        int_end,\n",
    "                        peak_begins[peak_i],\n",
    "                        peak_ends[peak_i]\n",
    "                    ))\n",
    "                if peak_i < len(peak_begins)-1:\n",
    "                    peak_i += 1\n",
    "                if debug:\n",
    "                    print('New peak: ({}, {})'.format(\n",
    "                        peak_begins[peak_i],\n",
    "                        peak_ends[peak_i]\n",
    "                    ))\n",
    "                ints_outside_peaks.append(interval_id)\n",
    "            elif int_begin <= peak_ends[peak_i]:\n",
    "                #   The beginning of interval is within\n",
    "                #   the peak:\n",
    "                #\n",
    "                #   interval: ------[-----]----\n",
    "                #   peak:     ---[-----]------\n",
    "                #\n",
    "                #   interval: ------[-----]----\n",
    "                #   peak:     ---[----------]--\n",
    "                if debug:\n",
    "                    print('Interval ({}, {}) overlaps current peak ({}, {})'.format(\n",
    "                        int_begin,\n",
    "                        int_end,\n",
    "                        peak_begins[peak_i],\n",
    "                        peak_ends[peak_i]\n",
    "                    ))\n",
    "                peak_id = chrm_to_peaks[chrm].iloc[peak_i]['peak_ids']\n",
    "                ints_in_peaks.append((interval_id, peak_id))\n",
    "        elif int_begin < peak_begins[peak_i]:\n",
    "            if int_end >= peak_begins[peak_i]:\n",
    "                #   The interval overlaps the beginning \n",
    "                #   of the peak:\n",
    "                #\n",
    "                #   interval: -[-----]--------\n",
    "                #   peak:     ---[-----]------\n",
    "                #\n",
    "                #   interval: -[-------------]-\n",
    "                #   peak:     ---[---------]---\n",
    "                # \n",
    "                if debug:\n",
    "                    print('Interval ({}, {}) overlaps current peak ({}, {})'.format(\n",
    "                        int_begin,\n",
    "                        int_end,\n",
    "                        peak_begins[peak_i],\n",
    "                        peak_ends[peak_i]\n",
    "                    ))\n",
    "                peak_id = chrm_to_peaks[chrm].iloc[peak_i]['peak_ids']\n",
    "                ints_in_peaks.append((interval_id, peak_id))\n",
    "            elif int_end < peak_begins[peak_i]:\n",
    "                #   The interval preceeds the peak:\n",
    "                #\n",
    "                #   interval: --[---]-----------\n",
    "                #   peak:     ----------[----]--\n",
    "                ints_outside_peaks.append(interval_id)\n",
    " \n",
    "print('{} intervals in a peak'.format(len(ints_in_peaks)))\n",
    "print('{} intervals outside a peak'.format(len(ints_outside_peaks)))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('chrX.12600_12800', 'Peak_220375,Peak_270645'),\n",
       " ('chrX.12800_13000', 'Peak_220375,Peak_270645'),\n",
       " ('chrX.13000_13200', 'Peak_220375,Peak_270645'),\n",
       " ('chrX.14000_14200', 'Peak_36870'),\n",
       " ('chrX.14200_14400', 'Peak_36870'),\n",
       " ('chrX.14400_14600', 'Peak_36870'),\n",
       " ('chrX.18800_19000',\n",
       "  'Peak_107552,Peak_249828,Peak_6451,Peak_64840,Peak_94523'),\n",
       " ('chrX.19000_19200',\n",
       "  'Peak_107552,Peak_249828,Peak_6451,Peak_64840,Peak_94523'),\n",
       " ('chrX.19200_19400',\n",
       "  'Peak_107552,Peak_249828,Peak_6451,Peak_64840,Peak_94523'),\n",
       " ('chrX.19400_19600',\n",
       "  'Peak_107552,Peak_249828,Peak_6451,Peak_64840,Peak_94523')]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ints_in_peaks[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>Peak-Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Within Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Within Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Within Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Within Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Within Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193533</th>\n",
       "      <td>32</td>\n",
       "      <td>Outside Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193534</th>\n",
       "      <td>43</td>\n",
       "      <td>Outside Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193535</th>\n",
       "      <td>40</td>\n",
       "      <td>Outside Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193536</th>\n",
       "      <td>20</td>\n",
       "      <td>Outside Peak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193537</th>\n",
       "      <td>20</td>\n",
       "      <td>Outside Peak</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7193538 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Count   Peak-Status\n",
       "0            0   Within Peak\n",
       "1            0   Within Peak\n",
       "2            0   Within Peak\n",
       "3            0   Within Peak\n",
       "4            0   Within Peak\n",
       "...        ...           ...\n",
       "7193533     32  Outside Peak\n",
       "7193534     43  Outside Peak\n",
       "7193535     40  Outside Peak\n",
       "7193536     20  Outside Peak\n",
       "7193537     20  Outside Peak\n",
       "\n",
       "[7193538 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_peak_counts = rnaseq_df.set_index('interval_id').loc[[x[0] for x in ints_in_peaks]]['n_reads']\n",
    "out_peak_counts = rnaseq_df.set_index('interval_id').loc[ints_outside_peaks]['n_reads']\n",
    "da = [\n",
    "    (x, 'Within Peak')\n",
    "    for x in in_peak_counts\n",
    "]\n",
    "da += [\n",
    "    (x, 'Outside Peak')\n",
    "    for x in out_peak_counts\n",
    "]\n",
    "df_in_out_peak_count = pd.DataFrame(\n",
    "    data=da,\n",
    "    columns=['Count', 'Peak-Status']\n",
    ")\n",
    "df_in_out_peak_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at the distribution of counts for intervals within a peak vs. outside a peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0, 20.0)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAai0lEQVR4nO3de5xdZX3v8c83wy2EcJM5EQci6qRQQAgwgFz0BC8IKXKxqHA8AhUa8Ega2qqlek6h6uHQl1UPBo80BYr2ULQKaCqREBHkKjAJgYRLzBQJZIgkkpoACZeEX/9YzyQ7O8+ebJLZe+3J/r5fr3nNWs961tq/7Kzkt551+S1FBGZmZtVGlB2AmZm1JicIMzPLcoIwM7MsJwgzM8tygjAzsywnCDMzy2pYgpC0t6Q7JD0u6TFJU1L77pJmSVqYfu9WY/2zU5+Fks5uVJxmZpanRj0HIWlPYM+ImCNpNDAbOBU4B1geEZdLuhjYLSL+qmrd3YFeoAeItO5hEfEfDQnWzMw20rARREQsiYg5afpF4AmgCzgF+G7q9l2KpFHtw8CsiFieksIs4IRGxWpmZhvbphkfImkf4BDgAWBMRCxJi34LjMms0gU8WzG/OLXltj0JmAQwatSow/bbb78hiro9LViwYKO2fffdt4RIzDbm/XPozZ49+3cR0Zlb1vAEIWkn4EbgoohYKWndsogISVt0jisipgHTAHp6eqK3t3dLNtf2JkyYsFHbnXfe2fQ4zHK8fw49SYtqLWvoXUyStqVIDtdHxE2p+fl0fWLgOsXSzKr9wN4V83ulNjMza5JG3sUk4BrgiYj4RsWi6cDAXUlnAz/JrD4TOF7Sbukup+NTmzVY9dGYj86slXj/bK5GjiCOAT4FvF/S3PQzEbgc+JCkhcAH0zySeiRdDRARy4GvAA+lny+nNjMza5KG3eZaBl+DMDN7cyTNjoie3DI/SW1mZllNuc3VhpfKO0V8jtdajffP5vEIwszMspwgbAPV95nn7js3K4v3z+ZygjAzsywnCDMzy3KCMDOzLCcIMzPLcoKwDbiUgbUy75/N5QRhZmZZflDONuKjMmtl3j+bxyMIMzPL8gjCzIYVl9poHo8gzMwsywnCzIYNl9poLp9iajFTp06lr6+v1Bj6+4u3u3Z1dZUaB0B3dzeTJ08uOwyztuQEYRtZvXp12SGYWQtwgmgxrXC0PGXKFACuuOKKkiMxszI1LEFIuhY4CVgaEQemth8A+6YuuwK/j4jxmXWfBl4E1gJrar0Oz8zMGqeRF6mvA06obIiIT0TE+JQUbgRuGmT941JfJwczA1xqo9kaNoKIiLsk7ZNbJknAx4H3N+rzzcxsy5R1DeK9wPMRsbDG8gBukxTAP0TEtOaFZmatzKOG5ikrQZwJ3DDI8mMjol/SfwFmSXoyIu7KdZQ0CZgEMHbs2KGP1MysTTU9QUjaBvgocFitPhHRn34vlXQzcASQTRBpdDENoKenJ4Y8YDNrKS610TxlPEn9QeDJiFicWyhplKTRA9PA8cD8JsZnZmY0MEFIugG4H9hX0mJJ56ZFZ1B1eknS2yTNSLNjgHskPQI8CNwSEbc2Kk4zGz5caqO5GnkX05k12s/JtD0HTEzTTwEHNyouMzOrj4v1mZlZlhOEmZllOUGYmVmWE4SZDRsutdFcThBmZpblct9mNqx41NA8HkGYmVmWRxBmNqy41EbzeARhZmZZThBmNmy41EZzOUGYmVmWE4SZmWU5QZiZWZYThJmZZTlBmNmw4VIbzeUEYWZmWX5QzsyGFY8amscjCDMzy/IIwsyGFZfaaJ6GjSAkXStpqaT5FW2XSuqXNDf9TKyx7gmSFkjqk3Rxo2I0M7PaGnmK6TrghEz7NyNifPqZUb1QUgfwbeBEYH/gTEn7NzBOMxsmXGqjuRqWICLiLmD5Zqx6BNAXEU9FxGvA94FThjQ4MzPbpDIuUl8o6dF0Cmq3zPIu4NmK+cWpLUvSJEm9knqXLVs21LGambWtZieI7wDvAsYDS4Cvb+kGI2JaRPRERE9nZ+eWbs7MzJKmJoiIeD4i1kbEG8A/UpxOqtYP7F0xv1dqMzOzJmpqgpC0Z8XsacD8TLeHgHGS3iFpO+AMYHoz4jOz1uZSG83VsOcgJN0ATAD2kLQYuASYIGk8EMDTwPmp79uAqyNiYkSskXQhMBPoAK6NiMcaFaeZmeU1LEFExJmZ5mtq9H0OmFgxPwPY6BZYMzOPGprHpTbMzCzLpTbMbFhxqY3m8QjCzMyynCDMbNhwqY3mcoIwM7MsJwgzM8tygjAzsywnCDMzy3KCMLNhw6U2mssJwszMsvygnJkNKx41NI9HEGZmluUEYWZmWU4QZmaW5WsQZlaXqVOn0tfXV3YY9PcXL5js6qr5qvqm6O7uZvLkyaXG0GhOEGY2rKxevbrsENqGE4SZ1aVVjpanTJkCwBVXXFFyJFs/X4MwM7OshiUISddKWippfkXb1yQ9KelRSTdL2rXGuk9LmidprqTeRsVoZma1NXIEcR1wQlXbLODAiDgI+DXw14Osf1xEjI+IngbFZ2Zmg2hYgoiIu4DlVW23RcSaNPsrYK9Gfb6ZmW2ZMq9BfBr4WY1lAdwmabakSYNtRNIkSb2SepctWzbkQZqZtatSEoSkLwFrgOtrdDk2Ig4FTgQ+K+l9tbYVEdMioiciejo7OxsQrZlZe2p6gpB0DnAS8MmIiFyfiOhPv5cCNwNHNC1AMzMDmpwgJJ0AfAE4OSJW1egzStLogWngeGB+rq+ZmTVOI29zvQG4H9hX0mJJ5wJXAqOBWekW1qtS37dJmpFWHQPcI+kR4EHgloi4tVFxmplZXsOepI6IMzPN19To+xwwMU0/BRzcqLjMzKw+fpLazMyynCDMzCzLCcLMzLKcIMzMLMsJwszMspwgzMwsywnCzMyynCDMzCzLCcLMzLLqShCSjqmnzczMth71jiCm1tlmZmZbiUFrMUk6Cjga6JT0FxWLdgY6GhmYmZmVa1PF+rYDdkr9Rle0rwROb1RQZmZWvkETRET8EvilpOsiYlGTYjIzsxZQb7nv7SVNA/apXCci3t+IoMzMrHz1JogfAlcBVwNrGxeOmZm1inoTxJqI+E5DIzEzs5ZS722u/ybpf0jaU9LuAz8NjczMzEpVb4I4G/g8cB8wO/30bmolSddKWippfkXb7pJmSVqYfu9WY92zU5+Fks6uM04zMxsidSWIiHhH5ueddax6HXBCVdvFwO0RMQ64Pc1vII1OLgGOBI4ALqmVSMzMrDHqugYh6axce0R8b7D1IuIuSftUNZ8CTEjT3wXuBP6qqs+HgVkRsTx9/iyKRHNDPfGamdmWq/ci9eEV0zsAHwDmAIMmiBrGRMSSNP1bYEymTxfwbMX84tS2EUmTgEkAY8eO3YxwzMwsp64EERGTK+cl7Qp8f0s/PCJCUmzhNqYB0wB6enq2aFtmZrbe5pb7fhl4x2au+7ykPQHS76WZPv3A3hXze6U2MzNrknqvQfwbMHB03gH8IfCvm/mZ0ynuiro8/f5Jps9M4LKKC9PHA3+9mZ9nZmabod5rEH9fMb0GWBQRize1kqQbKC5I7yFpMcWdSZcD/yrpXGAR8PHUtwe4ICLOi4jlkr4CPJQ29eWBC9ZmZtYc9V6D+KWkMay/WL2wzvXOrLHoA5m+vcB5FfPXAtfW8zlmZjb06n2j3MeBB4GPURzxPyDJ5b7NzLZi9Z5i+hJweEQsBZDUCfwc+FGjAjMzs3LVexfTiIHkkLzwJtY1M7NhqN4RxK2SZrL+SeZPADMaE5KZmbWCTb2TupviyefPS/oocGxadD9wfaODMzOz8mxqBPF/Sc8fRMRNwE0Akt6dln2kgbGZmVmJNnUdYUxEzKtuTG37NCQiMzNrCZtKELsOsmzkEMZhZmYtZlMJolfSn1Y3SjqP4qVBZma2ldrUNYiLgJslfZL1CaEH2A44rYFxmZlZyQZNEBHxPHC0pOOAA1PzLRHxi4ZHZmZmpaq3FtMdwB0NjsXMzFqIn4Y2M7MsJwgzM8tygjAzs6x6azFt9aZOnUpfX1/ZYbSEge9hypQpJUfSGrq7u5k8efKmO5ptZZwgkr6+PubOf4K1O+5ediilG/Fa8XbZ2U89X3Ik5etY5RcZWvtygqiwdsfdWb3fxLLDsBYy8kkXLbb21fRrEJL2lTS34melpIuq+kyQtKKiz980O04zs3bX9BFERCwAxgNI6gD6gZszXe+OiJOaGJqZmVUo+y6mDwD/HhGLSo7DzMyqlJ0gzmD9W+qqHSXpEUk/k3RArQ1ImiSpV1LvsmXLGhOlmVkbKi1BSNoOOBn4YWbxHODtEXEwMBX4ca3tRMS0iOiJiJ7Ozs6GxGpm1o7KHEGcCMxJBQE3EBErI+KlND0D2FbSHs0O0MysnZWZIM6kxuklSW+VpDR9BEWcLzQxNjOztlfKcxCSRgEfAs6vaLsAICKuAk4HPiNpDbAaOCMiooxYzczaVSkJIiJeBt5S1XZVxfSVwJXNjsvMzNYr+y4mMzNrUU4QZmaW5QRhZmZZThBmZpblBGFmZllOEGZmluUEYWZmWU4QZmaW5QRhZmZZThBmZpbld1KbDQNTp06lr6+v7DBawsD3MGXKlJIjaQ3d3d1Mnjy5Idt2gjAbBvr6+lj42MOM3Wlt2aGUbrvXixMfry7qLTmS8j3zUkdDt+8EYTZMjN1pLV88dGXZYVgLuWzOzg3dvq9BmJlZlhOEmZllOUGYmVmWE4SZmWU5QZiZWVZpCULS05LmSZoraaP71VT4lqQ+SY9KOrSMOM3M2lXZt7keFxG/q7HsRGBc+jkS+E76bWZmTdDKp5hOAb4XhV8Bu0ras+ygzMzaRZkJIoDbJM2WNCmzvAt4tmJ+cWrbgKRJknol9S5btqxBoZqZtZ8yE8SxEXEoxamkz0p63+ZsJCKmRURPRPR0dnYObYRmZm2stAQREf3p91LgZuCIqi79wN4V83ulNjMza4JSEoSkUZJGD0wDxwPzq7pNB85KdzO9B1gREUuaHKqZWdsq6y6mMcDNkgZi+JeIuFXSBQARcRUwA5gI9AGrgD8pKVYzs7ZUSoKIiKeAgzPtV1VMB/DZZsZlZmbrtfJtrmZmViInCDMzy3KCMDOzLCcIMzPLcoIwM7MsJwgzM8tygjAzsywnCDMzy3KCMDOzLCcIMzPLcoIwM7MsJwgzM8tygjAzsywnCDMzy3KCMDOzrLJeGNRy+vv76Vi1gpFPzig7FGshHateoL9/TdlhmJXCIwgzM8vyCCLp6urit69uw+r9JpYdirWQkU/OoKtrTNlhmJWi6SMISXtLukPS45IekzQl02eCpBWS5qafv2l2nGZm7a6MEcQa4C8jYo6k0cBsSbMi4vGqfndHxEklxGdmZpQwgoiIJRExJ02/CDwBdDU7DjMzG1ypF6kl7QMcAjyQWXyUpEck/UzSAc2NzMzMSrtILWkn4EbgoohYWbV4DvD2iHhJ0kTgx8C4GtuZBEwCGDt2bOMCNjNrM6WMICRtS5Ecro+Im6qXR8TKiHgpTc8AtpW0R25bETEtInoioqezs7OhcZuZtZMy7mIScA3wRER8o0aft6Z+SDqCIs4XmhelmZmVcYrpGOBTwDxJc1PbF4GxABFxFXA68BlJa4DVwBkRESXEambWtpqeICLiHkCb6HMlcGVzIjJrff39/bz8YgeXzdm57FCshSx6sYNR/f0N275LbZiZWZZLbZgNA11dXby6ZglfPLT6hj9rZ5fN2Zntuxr3GJlHEGZmluUEYWZmWU4QZmaW5QRhZmZZThBmZpblBGFmZllOEGZmluUEYWZmWX5QrkLHquWMfHJGqTGMeGUleuP1UmNoJTFiW97YobzyEh2rlgOt8U7qZ14qt9TG86tG8MraQavktJUdOoIxO75RagzPvNSRfw/CEHGCSLq7u8sOAYD+/jWsXr267DBaxsiRI+nqKvM/6DEtsW+0Qgwd/f2M8L65TsfIkQ19irke42jsvqGtqUhqT09P9Pb2lh2GmdmwIWl2RPTklvkahJmZZTlBmJlZlhOEmZllOUGYmVmWE4SZmWU5QZiZWVYpCULSCZIWSOqTdHFm+faSfpCWPyBpnxLCNDNra01PEJI6gG8DJwL7A2dK2r+q27nAf0REN/BN4O+aG6WZmZUxgjgC6IuIpyLiNeD7wClVfU4BvpumfwR8QJKf8Tcza6IySm10Ac9WzC8GjqzVJyLWSFoBvAX4XfXGJE0CJqXZlyQtGPKI29MeZL5vsxbh/XPovL3WgmFfiykipgHTyo5jayOpt9bj92Zl8/7ZHGWcYuoH9q6Y3yu1ZftI2gbYBXihKdGZmRlQToJ4CBgn6R2StgPOAKZX9ZkOnJ2mTwd+EVtTVUEzs2Gg6aeY0jWFC4GZQAdwbUQ8JunLQG9ETAeuAf5ZUh+wnCKJWHP5tJ21Mu+fTbBVlfs2M7Oh4yepzcwsywnCzMyynCCGAUnflHRRxfxMSVdXzH9d0l9IOnmgdImkUyufUJd0p6SNbguU1CPpW28ynqclzZP0qKTbJL11M/5M+0ia/2bXs9YgaS9JP5G0UNK/S7oi3XSyqfW+WEefqzPVFZB0jqQr30SMEyStkDRX0hOSLql33artXCrpc5uz7nDnBDE83AscDSBpBMVDQgdULD8auC8ipkfE5antVIpSJoOKiN6I+LPNiOm4iDgI6AU2+Y/eth6pqsFNwI8jYhzwB8BOwP+uY/VN7isRcV5EPL5lUa5zd0SMB3qA/y7p0CHabltwghge7gOOStMHAPOBFyXtJml74A+BOQNHWJKOBk4GvpaOnt6V1v2YpAcl/VrSe2HdUdZP0/Slkq5No42nJNWTOO4CuiV1SPqapIfSyOL8tM2dJN0uaU4adVSXVUHSOyU9LOnwLfiOrHneD7wSEf8EEBFrgT8HPi1px+ojfUk/TfvZ5cDItE9eL2mUpFskPSJpvqRPpP7rRruS/iTtrw8Cx1Rss1PSjWl/e0jSMQwiIl4GZlPsq++SdKuk2ZLulrRf2uZHUnHQhyX9XNKY6u1I+lNJP5M0cgu/w2Fh2D9J3Q4i4jlJaySNpRgt3E9RjuQoYAUwLyJeUypXFRH3SZoO/DQifgSQlm0TEUdImghcAnww83H7AccBo4EFkr4TEa8PEt5JwDyKAosrIuLwlLTulXQbRcmU0yJipaQ9gF+l2Ehx7UtRj+uciHhk874ha7IDKP6zXSf9/T4DdNdaKSIulnRhOqJH0h8Dz0XEH6X5XSr7S9oT+FvgMIr9/A7g4bT4CuCbEXFP+ncxk+JAKUvSW4D3AF+huEX2gohYKOlI4P9RJL17gPdEREg6D/gC8JcV27gQ+BBwakS8Osj3s9Vwghg+7qNIDkcD36BIEEdT/MO5t85t3JR+zwb2qdHnlrTzvyppKTCGol5WtTskrQUeBf4ncDVwkKTT0/JdgHFp3cskvQ94I8U9cGTWCfwE+OgQnlKw4WMe8HVJf0dxMHN31fIjgTsjYhmApB9QnM6C4uBmf62v4bmzpJ0i4qWqbbxX0sMU+97lwCKKfzc/rFh3+/R7L+AHKTFtB/ymYjtnURzsnLqJA6atihPE8DFwHeLdFKeYnqU4ulkJ/FOd2xg46llL7b/7yiOjwfodFxHriqWl89KTI2JmZSdJ51AkgsMi4nVJTwM7pMUrgGeAYwEniOHjcYoKB+tI2hkYC/QBB7Hh6esdyIiIX6drAhOBr0q6PSK+XGcMIyiO9l/ZRL+7I+Kkqjh/PzCKqTIV+EZETJc0Abi0Ytk8YDxFEvlN9YpbK1+DGD7uozidszwi1kbEcmBXitNM92X6v0hxmqhZZgKfkbQtgKQ/kDSKYiSxNCWH49iwcuRrwGnAWZL+WxNjtS1zO7CjpLNg3Ttevg5cFxGrgKeB8ZJGSNqbosT/gNcr9pG3Aasi4v8DXwOqLyA/APxXSW9J63ysYtltwOSBGUnj6wk8IlYCv5H0sbSeJB2cFu/C+rpwZ1et+jBwPjA9xd0WnCCGj3kUdy/9qqptReWRfIXvA59PF9zelVk+1K6mOLKco+L21X+gGH1cD/RImkcxTH+ycqV08fAk4M8lndyEOG0Lpbpop1Hc9LAQ+DXwCuvvULqX4ij7ceBbwJyK1acBj0q6nmI0/KCkuRTXxL5a9TlLKI7i70/bfKJi8Z9R7FePSnocuOBN/BE+CZwr6RHgMda/j+ZSilNPs8mUEo+Ie4DPAbek62lbPZfaMDOzLI8gzMwsywnCzMyynCDMzCzLCcLMzLKcIMzMLMsJwtqOpLWpHtB8ST+UtONmbGNdDatN9Dsp3Wr8iKTHtb5G1QbVdgdZv65+Zo3gBGHtaHVEjI+IAyke1nsz99DXLT3cNQ34SEQcDBwC3JkWn0od1XbfRD+zIecEYe3ubooKn6NUVLJ9MB3xnwLr3ltxt4pqtHNUVMrdgKTDazyQOJriYcEXACLi1YhYoEy13VQl9KE00rhRRVXUXL/KSqd7pNIlSDogxT43PTw2rkHfl7UR12KytiVpG+BE4FbgS8AvIuLTknaleML358BS4EMR8Ur6T/cGincLDGzjaIoaPqdExDOV24+I5aly7SJJtwM/BW6oUW339xHxj2n6q8C5ETE106/WH+cC4IqIuF7Fi3s6tvwbsnbnBGHtaGQq7wDFCOIainpWJ2v9m8N2oCg+9xxwZar1s5b11UShKC89DTg+Ip7LfVBEnCfp3RTVRz9HUS76nEzXA1Ni2JXi5TszM30Gcz/wJUl7ATdFxMI3ub7ZRpwgrB2trq7mmarR/nFELKhqvxR4HjiY4pRsZfXQJRSJ5BCKRIKkmRTlzHsj4jyAiJgHzJP0zxQ1is7JxHQdRSnpR1IF3Ak1Yl/D+lPD66qkRsS/SHoA+CNghqTzI+IXtb4As3r4GoRZYSYwOSUKJB2S2ncBlkTEG8Cn2PDUze8p/kP+P6k8NBHx4XQB/DwVb9ObUNF/PMX7CGDjarujgSXpwvYnK9qr+z1N8QIdqCi5LemdwFMR8S2Kd2wcVN8f26w2JwizwleAbSkqjT6W5qF429jZqfLnfsDLlStFxPMU1Wi/reLtZJUEfEHSgnRK629ZP3qorrb7vyjKW9/LhhVvq/v9PUVZ9YcpqvsO+DgwP33OgcD3NutbMKvgaq5mZpblEYSZmWU5QZiZWZYThJmZZTlBmJlZlhOEmZllOUGYmVmWE4SZmWX9J4aNrkAZiWLXAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "df_in_out_peak_count = df_in_out_peak_count.loc[df_in_out_peak_count['Count']>0]\n",
    "\n",
    "sns.boxplot(data=df_in_out_peak_count, x='Peak-Status', y='Count')\n",
    "plt.ylim((0,20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at fraction of intervals with zero counts within a peak vs. outside a peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of zero counts within peak:  0.7699418350115704\n",
      "Fraction of zero counts outside peak:  0.9100182437335504\n"
     ]
    }
   ],
   "source": [
    "frac_zero_in_peak = len([x for x in in_peak_counts if x == 0])/len(in_peak_counts)\n",
    "frac_zero_out_peak = len([x for x in out_peak_counts if x == 0])/len(out_peak_counts)\n",
    "print(\"Fraction of zero counts within peak: \", frac_zero_in_peak)\n",
    "print(\"Fraction of zero counts outside peak: \", frac_zero_out_peak)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform a Wilcoxon Rank-Sum test to see if distribution of counts are different between the two"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p-value:  0.0\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import ranksums\n",
    "\n",
    "_, p_val = ranksums(in_peak_counts, out_peak_counts)\n",
    "print('p-value: ', p_val)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
